File: artifacts/api-server/src/routes/scheduled-interviews.ts
import { Router, type IRouter } from "express";
import { eq, and, inArray } from "drizzle-orm";
import {
  db,
  scheduledInterviewsTable,
  interviewCandidatesTable,
  interviewsTable,
  interviewReportsTable,
  jobsTable,
  usersTable,
} from "@workspace/db";
import ExcelJS from "exceljs";

const router: IRouter = Router();

router.get("/scheduled-interviews", async (req, res) => {
  if (!req.isAuthenticated()) { res.status(401).json({ error: "Unauthorized" }); return; }
  const list = await db.select().from(scheduledInterviewsTable)
    .where(eq(scheduledInterviewsTable.employerId, req.user.id));

  const withCounts = await Promise.all(list.map(async (si) => {
    const candidates = await db.select().from(interviewCandidatesTable)
      .where(eq(interviewCandidatesTable.scheduledInterviewId, si.id));
    const attempts = await db.select().from(interviewsTable)
      .where(eq(interviewsTable.scheduledInterviewId, si.id));
    let jobTitle: string | null = null;
    if (si.jobId) {
      const [j] = await db.select().from(jobsTable).where(eq(jobsTable.id, si.jobId));
      jobTitle = j?.title ?? null;
    }
    return { ...si, jobTitle, candidateCount: candidates.length, attemptCount: attempts.length };
  }));

  res.json(withCounts);
});

router.post("/scheduled-interviews", async (req, res) => {
  if (!req.isAuthenticated()) { res.status(401).json({ error: "Unauthorized" }); return; }
  const { jobId, title, startTime, deadlineTime, codingQuestionsCount, role, difficulty, interviewStyle } = req.body;
  if (!title || !startTime || !deadlineTime) {
    res.status(400).json({ error: "title, startTime, deadlineTime are required" }); return;
  }
  const [si] = await db.insert(scheduledInterviewsTable).values({
    employerId: req.user.id,
    jobId: jobId ?? null,
    title,
    startTime: new Date(startTime),
    deadlineTime: new Date(deadlineTime),
    codingQuestionsCount: codingQuestionsCount ?? 0,
    role: role ?? "Software Engineer",
    difficulty: difficulty ?? "Medium",
    interviewStyle: interviewStyle ?? "Friendly",
  }).returning();
  res.status(201).json(si);
});

router.get("/scheduled-interviews/:id", async (req, res) => {
  if (!req.isAuthenticated()) { res.status(401).json({ error: "Unauthorized" }); return; }
  const id = parseInt(req.params.id);
  const [si] = await db.select().from(scheduledInterviewsTable)
    .where(and(eq(scheduledInterviewsTable.id, id), eq(scheduledInterviewsTable.employerId, req.user.id)));
  if (!si) { res.status(404).json({ error: "Not found" }); return; }

  const candidates = await db.select().from(interviewCandidatesTable)
    .where(eq(interviewCandidatesTable.scheduledInterviewId, id));
  let jobTitle: string | null = null;
  if (si.jobId) {
    const [j] = await db.select().from(jobsTable).where(eq(jobsTable.id, si.jobId));
    jobTitle = j?.title ?? null;
  }
  res.json({ ...si, jobTitle, candidates });
});

router.delete("/scheduled-interviews/:id", async (req, res) => {
  if (!req.isAuthenticated()) { res.status(401).json({ error: "Unauthorized" }); return; }
  const id = parseInt(req.params.id);
  const [si] = await db.select().from(scheduledInterviewsTable)
    .where(and(eq(scheduledInterviewsTable.id, id), eq(scheduledInterviewsTable.employerId, req.user.id)));
  if (!si) { res.status(404).json({ error: "Not found" }); return; }
  await db.delete(scheduledInterviewsTable).where(eq(scheduledInterviewsTable.id, id));
  res.json({ success: true });
});

router.post("/scheduled-interviews/:id/candidates", async (req, res) => {
  if (!req.isAuthenticated()) { res.status(401).json({ error: "Unauthorized" }); return; }
  const id = parseInt(req.params.id);
  const [si] = await db.select().from(scheduledInterviewsTable)
    .where(and(eq(scheduledInterviewsTable.id, id), eq(scheduledInterviewsTable.employerId, req.user.id)));
  if (!si) { res.status(404).json({ error: "Not found" }); return; }

  const { emails } = req.body as { emails: string[] };
  if (!emails || !Array.isArray(emails)) { res.status(400).json({ error: "emails array required" }); return; }

  const existing = await db.select({ email: interviewCandidatesTable.email })
    .from(interviewCandidatesTable)
    .where(eq(interviewCandidatesTable.scheduledInterviewId, id));
  const existingSet = new Set(existing.map(e => e.email.toLowerCase()));
  const toAdd = [...new Set(emails.map(e => e.toLowerCase().trim()).filter(e => e && /\S+@\S+\.\S+/.test(e)))].filter(e => !existingSet.has(e));

  if (toAdd.length > 0) {
    await db.insert(interviewCandidatesTable).values(
      toAdd.map(email => ({ scheduledInterviewId: id, email }))
    );
  }
  const all = await db.select().from(interviewCandidatesTable)
    .where(eq(interviewCandidatesTable.scheduledInterviewId, id));
  res.json({ added: toAdd.length, candidates: all });
});

router.delete("/scheduled-interviews/:id/candidates", async (req, res) => {
  if (!req.isAuthenticated()) { res.status(401).json({ error: "Unauthorized" }); return; }
  const id = parseInt(req.params.id);
  const [si] = await db.select().from(scheduledInterviewsTable)
    .where(and(eq(scheduledInterviewsTable.id, id), eq(scheduledInterviewsTable.employerId, req.user.id)));
  if (!si) { res.status(404).json({ error: "Not found" }); return; }

  const { email } = req.body as { email: string };
  if (!email) { res.status(400).json({ error: "email required" }); return; }
  await db.delete(interviewCandidatesTable).where(
    and(
      eq(interviewCandidatesTable.scheduledInterviewId, id),
      eq(interviewCandidatesTable.email, email.toLowerCase().trim())
    )
  );
  const remaining = await db.select().from(interviewCandidatesTable)
    .where(eq(interviewCandidatesTable.scheduledInterviewId, id));
  res.json({ candidates: remaining });
});

router.get("/scheduled-interviews/:id/results", async (req, res) => {
  if (!req.isAuthenticated()) { res.status(401).json({ error: "Unauthorized" }); return; }
  const id = parseInt(req.params.id);
  const [si] = await db.select().from(scheduledInterviewsTable)
    .where(and(eq(scheduledInterviewsTable.id, id), eq(scheduledInterviewsTable.employerId, req.user.id)));
  if (!si) { res.status(404).json({ error: "Not found" }); return; }

  if (new Date() < new Date(si.startTime)) {
    res.status(403).json({ error: "Results not available until interview start time" }); return;
  }

  const candidates = await db.select().from(interviewCandidatesTable)
    .where(eq(interviewCandidatesTable.scheduledInterviewId, id));

  const attempts = await db.select().from(interviewsTable)
    .where(eq(interviewsTable.scheduledInterviewId, id));

  const results = await Promise.all(attempts.map(async (iv) => {
    const [report] = await db.select().from(interviewReportsTable)
      .where(eq(interviewReportsTable.interviewId, iv.id));
    const [user] = await db.select().from(usersTable).where(eq(usersTable.id, iv.userId));
    return {
      interviewId: iv.id,
      candidateName: iv.candidateName ?? user?.firstName ?? "Unknown",
      email: user?.email ?? "",
      status: iv.status,
      report: report ?? null,
      overallScore: report?.overallScore ?? null,
      recommendation: report?.recommendation ?? null,
      skillScores: report?.skillScores ?? [],
      englishScore: report?.englishScore ?? null,
      behavioralScore: report?.behavioralScore ?? null,
      confidenceScore: report?.confidenceScore ?? null,
    };
  }));

  const attemptedEmails = new Set(results.map(r => r.email.toLowerCase()));
  const notAttempted = candidates
    .filter(c => !attemptedEmails.has(c.email.toLowerCase()))
    .map(c => ({ email: c.email, candidateName: null, status: "not_attempted", report: null }));

  res.json({ scheduledInterview: si, attempted: results, notAttempted });
});

router.get("/scheduled-interviews/:id/results/export", async (req, res) => {
  if (!req.isAuthenticated()) { res.status(401).json({ error: "Unauthorized" }); return; }
  const id = parseInt(req.params.id);
  const [si] = await db.select().from(scheduledInterviewsTable)
    .where(and(eq(scheduledInterviewsTable.id, id), eq(scheduledInterviewsTable.employerId, req.user.id)));
  if (!si) { res.status(404).json({ error: "Not found" }); return; }

  const candidates = await db.select().from(interviewCandidatesTable)
    .where(eq(interviewCandidatesTable.scheduledInterviewId, id));
  const attempts = await db.select().from(interviewsTable)
    .where(eq(interviewsTable.scheduledInterviewId, id));

  const results = await Promise.all(attempts.map(async (iv) => {
    const [report] = await db.select().from(interviewReportsTable)
      .where(eq(interviewReportsTable.interviewId, iv.id));
    const [user] = await db.select().from(usersTable).where(eq(usersTable.id, iv.userId));
    return {
      candidateName: iv.candidateName ?? user?.firstName ?? "Unknown",
      email: user?.email ?? "",
      overallScore: report?.overallScore ?? null,
      englishScore: report?.englishScore ?? null,
      behavioralScore: report?.behavioralScore ?? null,
      confidenceScore: report?.confidenceScore ?? null,
      skillScores: report?.skillScores ?? [],
      recommendation: report?.recommendation ?? null,
      feedback: report?.feedback ?? "",
    };
  }));

  const attemptedEmails = new Set(results.map(r => r.email.toLowerCase()));
  const notAttempted = candidates
    .filter(c => !attemptedEmails.has(c.email.toLowerCase()))
    .map(c => ({ candidateName: "", email: c.email, overallScore: null, englishScore: null, behavioralScore: null, confidenceScore: null, skillScores: [], recommendation: null, feedback: "" }));

  const workbook = new ExcelJS.Workbook();
  workbook.creator = "Vocalize.ai";

  const baseHeaders = ["Name", "Email", "Overall Score", "English Score", "Behavioral Score", "Confidence Score"];

  const allSkillNames = [...new Set(results.flatMap(r => r.skillScores.map((s: any) => s.skill)))];
  const skillHeaders = allSkillNames.map(s => `Skill: ${s}`);
  const headers = [...baseHeaders, ...skillHeaders, "Recommendation", "Feedback"];

  const makeRow = (r: typeof results[0]) => {
    const base = [r.candidateName, r.email, r.overallScore, r.englishScore, r.behavioralScore, r.confidenceScore];
    const skills = allSkillNames.map(sn => {
      const found = r.skillScores.find((s: any) => s.skill === sn);
      return found ? (found as any).score : "";
    });
    return [...base, ...skills, r.recommendation, r.feedback];
  };

  const makeNotAttemptedRow = (r: typeof notAttempted[0]) => {
    return [r.candidateName, r.email, ...Array(headers.length - 2).fill("")];
  };

  const strongly = results.filter(r => r.recommendation === "hire" && (r.overallScore ?? 0) >= 80);
  const hire = results.filter(r => r.recommendation === "hire" && (r.overallScore ?? 0) < 80);
  const maybe = results.filter(r => r.recommendation === "maybe");
  const noHire = results.filter(r => r.recommendation === "no_hire");

  const sheets: { name: string; rows: any[][] }[] = [
    { name: "Strongly Hire", rows: strongly.map(makeRow) },
    { name: "Hire", rows: [...hire, ...maybe].map(makeRow) },
    { name: "Do Not Hire", rows: noHire.map(makeRow) },
    { name: "Not Attempted", rows: notAttempted.map(makeNotAttemptedRow) },
  ];

  for (const sheet of sheets) {
    const ws = workbook.addWorksheet(sheet.name);
    const notAttemptedHeaders = ["Name", "Email"];
    ws.addRow(sheet.name === "Not Attempted" ? notAttemptedHeaders : headers);
    ws.getRow(1).font = { bold: true };
    ws.getRow(1).fill = { type: "pattern", pattern: "solid", fgColor: { argb: "FF4F46E5" } };
    ws.getRow(1).font = { bold: true, color: { argb: "FFFFFFFF" } };
    for (const row of sheet.rows) ws.addRow(row);
    ws.columns.forEach(col => { col.width = 18; });
  }

  const buf = await workbook.xlsx.writeBuffer();
  res.setHeader("Content-Disposition", `attachment; filename="results-${id}.xlsx"`);
  res.setHeader("Content-Type", "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet");
  res.send(buf);
});

router.post("/scheduled-interviews/validate-access", async (req, res) => {
  const { scheduledInterviewId, email } = req.body;
  if (!scheduledInterviewId || !email) {
    res.status(400).json({ error: "scheduledInterviewId and email required" }); return;
  }
  const [si] = await db.select().from(scheduledInterviewsTable)
    .where(eq(scheduledInterviewsTable.id, parseInt(scheduledInterviewId)));
  if (!si) { res.status(404).json({ allowed: false, reason: "Interview not found" }); return; }

  const now = new Date();
  if (now < new Date(si.startTime)) {
    res.json({ allowed: false, reason: "Interview has not started yet", startTime: si.startTime }); return;
  }
  if (now > new Date(si.deadlineTime)) {
    res.json({ allowed: false, reason: "Interview deadline has passed" }); return;
  }

  const [candidate] = await db.select().from(interviewCandidatesTable)
    .where(and(
      eq(interviewCandidatesTable.scheduledInterviewId, parseInt(scheduledInterviewId)),
      eq(interviewCandidatesTable.email, email.toLowerCase().trim())
    ));

  if (!candidate) {
    res.json({ allowed: false, reason: "You are not registered for this interview" }); return;
  }

  res.json({ allowed: true, scheduledInterview: si });
});

export default router;

File: artifacts/api-server-py/routes/scheduled_interviews.py
import io
from datetime import datetime
from fastapi import APIRouter, Request, Depends, HTTPException
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session as DBSession
from db.database import get_db
from db.models import ScheduledInterview, InterviewCandidate, Interview, InterviewReport, Job, User
from routes.auth import get_current_user

router = APIRouter()


@router.get("/scheduled-interviews")
async def list_scheduled(request: Request, db: DBSession = Depends(get_db)):
    user = await get_current_user(request, db)
    if not user:
        raise HTTPException(status_code=401, detail="Unauthorized")
    items = db.query(ScheduledInterview).filter(ScheduledInterview.employer_id == user["id"]).all()
    result = []
    for si in items:
        candidates = db.query(InterviewCandidate).filter(InterviewCandidate.scheduled_interview_id == si.id).all()
        attempts = db.query(Interview).filter(Interview.scheduled_interview_id == si.id).all()
        job_title = None
        if si.job_id:
            j = db.query(Job).filter(Job.id == si.job_id).first()
            job_title = j.title if j else None
        d = _si_to_dict(si)
        d["jobTitle"] = job_title
        d["candidateCount"] = len(candidates)
        d["attemptCount"] = len(attempts)
        result.append(d)
    return result


@router.post("/scheduled-interviews", status_code=201)
async def create_scheduled(request: Request, db: DBSession = Depends(get_db)):
    user = await get_current_user(request, db)
    if not user:
        raise HTTPException(status_code=401, detail="Unauthorized")
    body = await request.json()
    title = body.get("title")
    start_time = body.get("startTime")
    deadline_time = body.get("deadlineTime")
    if not title or not start_time or not deadline_time:
        raise HTTPException(status_code=400, detail="title, startTime, deadlineTime are required")

    si = ScheduledInterview(
        employer_id=user["id"],
        job_id=body.get("jobId"),
        title=title,
        start_time=datetime.fromisoformat(start_time.replace("Z", "+00:00")),
        deadline_time=datetime.fromisoformat(deadline_time.replace("Z", "+00:00")),
        coding_questions_count=body.get("codingQuestionsCount", 0),
        role=body.get("role", "Software Engineer"),
        difficulty=body.get("difficulty", "Medium"),
        interview_style=body.get("interviewStyle", "Professional"),
        interview_type=body.get("interviewType", "Mixed"),
        coding_language=body.get("codingLanguage", "Candidate's Choice"),
        question_complexity=body.get("questionComplexity", "Moderate"),
    )
    db.add(si)
    db.commit()
    db.refresh(si)
    return _si_to_dict(si)


@router.get("/scheduled-interviews/{si_id}")
async def get_scheduled(si_id: int, request: Request, db: DBSession = Depends(get_db)):
    user = await get_current_user(request, db)
    if not user:
        raise HTTPException(status_code=401, detail="Unauthorized")
    si = db.query(ScheduledInterview).filter(ScheduledInterview.id == si_id, ScheduledInterview.employer_id == user["id"]).first()
    if not si:
        raise HTTPException(status_code=404, detail="Not found")
    candidates = db.query(InterviewCandidate).filter(InterviewCandidate.scheduled_interview_id == si_id).all()
    job_title = None
    if si.job_id:
        j = db.query(Job).filter(Job.id == si.job_id).first()
        job_title = j.title if j else None
    result = _si_to_dict(si)
    result["jobTitle"] = job_title
    result["candidates"] = [_candidate_to_dict(c) for c in candidates]
    return result


@router.delete("/scheduled-interviews/{si_id}")
async def delete_scheduled(si_id: int, request: Request, db: DBSession = Depends(get_db)):
    user = await get_current_user(request, db)
    if not user:
        raise HTTPException(status_code=401, detail="Unauthorized")
    si = db.query(ScheduledInterview).filter(ScheduledInterview.id == si_id, ScheduledInterview.employer_id == user["id"]).first()
    if not si:
        raise HTTPException(status_code=404, detail="Not found")
    db.delete(si)
    db.commit()
    return {"success": True}


@router.post("/scheduled-interviews/{si_id}/candidates")
async def add_candidates(si_id: int, request: Request, db: DBSession = Depends(get_db)):
    user = await get_current_user(request, db)
    if not user:
        raise HTTPException(status_code=401, detail="Unauthorized")
    si = db.query(ScheduledInterview).filter(ScheduledInterview.id == si_id, ScheduledInterview.employer_id == user["id"]).first()
    if not si:
        raise HTTPException(status_code=404, detail="Not found")
    body = await request.json()
    emails = body.get("emails", [])
    if not isinstance(emails, list):
        raise HTTPException(status_code=400, detail="emails array required")

    existing = {c.email.lower() for c in db.query(InterviewCandidate).filter(InterviewCandidate.scheduled_interview_id == si_id).all()}
    import re
    to_add = list({e.lower().strip() for e in emails if e and re.match(r"\S+@\S+\.\S+", e)} - existing)

    for email in to_add:
        db.add(InterviewCandidate(scheduled_interview_id=si_id, email=email))
    db.commit()

    all_candidates = db.query(InterviewCandidate).filter(InterviewCandidate.scheduled_interview_id == si_id).all()
    return {"added": len(to_add), "candidates": [_candidate_to_dict(c) for c in all_candidates]}


@router.delete("/scheduled-interviews/{si_id}/candidates")
async def remove_candidate(si_id: int, request: Request, db: DBSession = Depends(get_db)):
    user = await get_current_user(request, db)
    if not user:
        raise HTTPException(status_code=401, detail="Unauthorized")
    si = db.query(ScheduledInterview).filter(ScheduledInterview.id == si_id, ScheduledInterview.employer_id == user["id"]).first()
    if not si:
        raise HTTPException(status_code=404, detail="Not found")
    body = await request.json()
    email = body.get("email", "").lower().strip()
    if not email:
        raise HTTPException(status_code=400, detail="email required")
    c = db.query(InterviewCandidate).filter(
        InterviewCandidate.scheduled_interview_id == si_id,
        InterviewCandidate.email == email
    ).first()
    if c:
        db.delete(c)
        db.commit()
    remaining = db.query(InterviewCandidate).filter(InterviewCandidate.scheduled_interview_id == si_id).all()
    return {"candidates": [_candidate_to_dict(c) for c in remaining]}


@router.get("/scheduled-interviews/{si_id}/results")
async def get_results(si_id: int, request: Request, db: DBSession = Depends(get_db)):
    user = await get_current_user(request, db)
    if not user:
        raise HTTPException(status_code=401, detail="Unauthorized")
    si = db.query(ScheduledInterview).filter(ScheduledInterview.id == si_id, ScheduledInterview.employer_id == user["id"]).first()
    if not si:
        raise HTTPException(status_code=404, detail="Not found")
    if datetime.utcnow() < si.start_time:
        raise HTTPException(status_code=403, detail="Results not available until interview start time")

    candidates = db.query(InterviewCandidate).filter(InterviewCandidate.scheduled_interview_id == si_id).all()
    attempts = db.query(Interview).filter(Interview.scheduled_interview_id == si_id).all()

    results = []
    for iv in attempts:
        report = db.query(InterviewReport).filter(InterviewReport.interview_id == iv.id).first()
        db_user = db.query(User).filter(User.id == iv.user_id).first()
        results.append({
            "interviewId": iv.id,
            "candidateName": iv.candidate_name or (db_user.first_name if db_user else "Unknown"),
            "email": db_user.email if db_user else "",
            "status": iv.status,
            "report": _report_to_dict(report) if report else None,
            "overallScore": report.overall_score if report else None,
            "recommendation": report.recommendation if report else None,
            "skillScores": report.skill_scores if report else [],
            "englishScore": report.english_score if report else None,
            "behavioralScore": report.behavioral_score if report else None,
            "confidenceScore": report.confidence_score if report else None,
            "codingScore": getattr(report, "coding_score", None) if report else None,
            "technicalScore": getattr(report, "technical_score", None) if report else None,
        })

    attempted_emails = {r["email"].lower() for r in results}
    not_attempted = [
        {"email": c.email, "candidateName": None, "status": "not_attempted", "report": None}
        for c in candidates if c.email.lower() not in attempted_emails
    ]

    return {"scheduledInterview": _si_to_dict(si), "attempted": results, "notAttempted": not_attempted}


@router.get("/scheduled-interviews/{si_id}/results/export")
async def export_results(si_id: int, request: Request, db: DBSession = Depends(get_db)):
    user = await get_current_user(request, db)
    if not user:
        raise HTTPException(status_code=401, detail="Unauthorized")
    si = db.query(ScheduledInterview).filter(ScheduledInterview.id == si_id, ScheduledInterview.employer_id == user["id"]).first()
    if not si:
        raise HTTPException(status_code=404, detail="Not found")

    from openpyxl import Workbook
    from openpyxl.styles import Font, PatternFill

    candidates = db.query(InterviewCandidate).filter(InterviewCandidate.scheduled_interview_id == si_id).all()
    attempts = db.query(Interview).filter(Interview.scheduled_interview_id == si_id).all()

    results = []
    for iv in attempts:
        report = db.query(InterviewReport).filter(InterviewReport.interview_id == iv.id).first()
        db_user = db.query(User).filter(User.id == iv.user_id).first()
        skill_scores = report.skill_scores if report else []
        results.append({
            "candidateName": iv.candidate_name or (db_user.first_name if db_user else "Unknown"),
            "email": db_user.email if db_user else "",
            "overallScore": report.overall_score if report else None,
            "englishScore": report.english_score if report else None,
            "behavioralScore": report.behavioral_score if report else None,
            "confidenceScore": report.confidence_score if report else None,
            "codingScore": getattr(report, "coding_score", None) if report else None,
            "technicalScore": getattr(report, "technical_score", None) if report else None,
            "skillScores": skill_scores,
            "recommendation": report.recommendation if report else None,
            "feedback": report.feedback if report else "",
        })

    all_skill_names = list({s["skill"] for r in results for s in r["skillScores"]})
    base_headers = ["Name", "Email", "Overall Score", "Behavioral Score", "Coding Score", "Technical Theory Score", "English Score", "Confidence Score"]
    skill_headers = [f"Skill: {s}" for s in all_skill_names]
    headers = base_headers + skill_headers + ["Recommendation", "Feedback"]

    wb = Workbook()
    wb.remove(wb.active)

    header_font = Font(bold=True, color="FFFFFF")
    header_fill = PatternFill(start_color="4F46E5", end_color="4F46E5", fill_type="solid")

    attempted_emails = {r["email"].lower() for r in results}
    not_attempted = [{"candidateName": "", "email": c.email} for c in candidates if c.email.lower() not in attempted_emails]

    strongly_hire = [r for r in results if r["recommendation"] == "hire" and (r["overallScore"] or 0) >= 80]
    hire = [r for r in results if r["recommendation"] == "hire" and (r["overallScore"] or 0) < 80]
    maybe = [r for r in results if r["recommendation"] == "maybe"]
    no_hire = [r for r in results if r["recommendation"] == "no_hire"]

    def add_sheet(name, rows, use_full_headers=True):
        ws = wb.create_sheet(title=name)
        hdrs = headers if use_full_headers else ["Name", "Email"]
        ws.append(hdrs)
        for cell in ws[1]:
            cell.font = header_font
            cell.fill = header_fill
        for r in rows:
            ws.append(r)
        for col in ws.columns:
            col_letter = col[0].column_letter
            ws.column_dimensions[col_letter].width = 18

    def make_row(r):
        base = [r["candidateName"], r["email"], r["overallScore"], r["behavioralScore"], r["codingScore"], r["technicalScore"], r["englishScore"], r["confidenceScore"]]
        skills = [next((s["score"] for s in r["skillScores"] if s["skill"] == sn), "") for sn in all_skill_names]
        return base + skills + [r["recommendation"], r["feedback"]]

    add_sheet("Strongly Hire", [make_row(r) for r in strongly_hire])
    add_sheet("Hire", [make_row(r) for r in hire + maybe])
    add_sheet("Do Not Hire", [make_row(r) for r in no_hire])
    add_sheet("Not Attempted", [[r["candidateName"], r["email"]] for r in not_attempted], use_full_headers=False)

    buffer = io.BytesIO()
    wb.save(buffer)
    buffer.seek(0)

    return StreamingResponse(
        buffer,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": f"attachment; filename=results-{si_id}.xlsx"}
    )


@router.post("/scheduled-interviews/validate-access")
async def validate_access(request: Request, db: DBSession = Depends(get_db)):
    body = await request.json()
    si_id = body.get("scheduledInterviewId")
    email = body.get("email", "").lower().strip()
    if not si_id or not email:
        raise HTTPException(status_code=400, detail="scheduledInterviewId and email required")

    si = db.query(ScheduledInterview).filter(ScheduledInterview.id == int(si_id)).first()
    if not si:
        return {"allowed": False, "reason": "Interview not found"}

    now = datetime.utcnow()
    if now < si.start_time:
        return {"allowed": False, "reason": "Interview has not started yet", "startTime": si.start_time.isoformat()}
    if now > si.deadline_time:
        return {"allowed": False, "reason": "Interview deadline has passed"}

    candidate = db.query(InterviewCandidate).filter(
        InterviewCandidate.scheduled_interview_id == int(si_id),
        InterviewCandidate.email == email
    ).first()

    if not candidate:
        return {"allowed": False, "reason": "You are not registered for this interview"}

    return {"allowed": True, "scheduledInterview": _si_to_dict(si)}


def _si_to_dict(si: ScheduledInterview) -> dict:
    return {
        "id": si.id,
        "employerId": si.employer_id,
        "jobId": si.job_id,
        "title": si.title,
        "startTime": si.start_time.isoformat() if si.start_time else None,
        "deadlineTime": si.deadline_time.isoformat() if si.deadline_time else None,
        "codingQuestionsCount": si.coding_questions_count,
        "role": si.role,
        "difficulty": si.difficulty,
        "interviewStyle": si.interview_style,
        "interviewType": getattr(si, "interview_type", "Mixed"),
        "codingLanguage": getattr(si, "coding_language", "Candidate's Choice"),
        "questionComplexity": getattr(si, "question_complexity", "Moderate"),
        "createdAt": si.created_at.isoformat() if si.created_at else None,
    }


def _candidate_to_dict(c: InterviewCandidate) -> dict:
    return {
        "id": c.id,
        "scheduledInterviewId": c.scheduled_interview_id,
        "email": c.email,
        "addedAt": c.added_at.isoformat() if c.added_at else None,
    }


def _report_to_dict(r: InterviewReport) -> dict:
    return {
        "id": r.id,
        "interviewId": r.interview_id,
        "englishScore": r.english_score,
        "englishFeedback": r.english_feedback,
        "overallScore": r.overall_score,
        "confidenceScore": r.confidence_score,
        "confidenceNotes": r.confidence_notes,
        "behavioralScore": r.behavioral_score,
        "behavioralAnalysis": r.behavioral_analysis,
        "communicationAnalysis": r.communication_analysis,
        "answerQualityBreakdown": r.answer_quality_breakdown,
        "stutterAnalysis": r.stutter_analysis,
        "skillScores": r.skill_scores,
        "recommendation": r.recommendation,
        "feedback": r.feedback,
        "createdAt": r.created_at.isoformat() if r.created_at else None,
    }
